src/data/matrix/block.lean

```diff
@@ -26,6 +26,22 @@ open_locale matrix
 
 namespace matrix
 
+section vector
+
+open_locale big_operators
+
+lemma dot_product_block' [has_mul α] [add_comm_monoid α] (v w : m ⊕ n → α) :
+dot_product v w = ∑ i, v (sum.inl i) * w (sum.inl i) + ∑ j, v (sum.inr j) * w (sum.inr j) :=
+by {rw [dot_product, ←fintype.sum_sum_elim], congr, ext (i | j); simp }
```

Comment:
Can you get there faster with `fintype.sum_sum_type`?

Reply:
Sure. I will have a look.